src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java

```diff
@@ -110,13 +124,31 @@ private void updateOriginalQueryResults(final List<QuerySearchResult> querySearc
         for (int index = 0; index < querySearchResults.size(); index++) {
             QuerySearchResult querySearchResult = querySearchResults.get(index);
             CompoundTopDocs updatedTopDocs = queryTopDocs.get(index);
-            float maxScore = updatedTopDocs.getTotalHits().value > 0 ? updatedTopDocs.getScoreDocs().get(0).score : 0.0f;
+            float maxScore = 0.0f;
+            if (updatedTopDocs.getTotalHits().value > 0 && !updatedTopDocs.getScoreDocs().isEmpty()) {
+                maxScore = updatedTopDocs.getScoreDocs().get(0).score;
+            }
+            TopDocsAndMaxScore updatedTopDocsAndMaxScore;
+            if (sort == null) {
+                // create final version of top docs with all updated values
+                TopDocs topDocs = new TopDocs(updatedTopDocs.getTotalHits(), updatedTopDocs.getScoreDocs().toArray(new ScoreDoc[0]));
+                updatedTopDocsAndMaxScore = new TopDocsAndMaxScore(topDocs, maxScore);
+            } else {
+                final FieldDoc[] fieldDocs = new FieldDoc[updatedTopDocs.getScoreDocs().size()];
+                int i = 0;
+                if (updatedTopDocs.getTotalHits().value > 0) {
+                    for (ScoreDoc scoreDoc : updatedTopDocs.getScoreDocs()) {
+                        maxScore = Math.max(maxScore, scoreDoc.score);
+                        fieldDocs[i++] = (FieldDoc) scoreDoc;
+                    }
+                } else {
+                    maxScore = MAX_SCORE_WHEN_NO_HITS_FOUND;
```

Comment:
can you add comment on why we have to update maxscore if sort != null and total hits > 0 ?